### PR TITLE
307 use ce lib

### PR DESCRIPTION
**Pull request by mmacai**
## Description

Based on the https://github.com/Accenture/reactive-interaction-gateway/pull/306

TODO:

- [x] passing integration tests
- [x] passing smoke tests
- [x] passing unit tests
- [x] passing cypress tests
- [x] changelog entry (point out breaking changes)
- [x] docs update
- [x] ce lib new release
- [x] cleanup

## What to look out for

- incorporated ce lib to use kafka transport protocol
- kafka consumer should properly use binary/structured mode
- kafka to proxy shouldn't merge headers with body as a HTTP response
- fixed some npm audit vulnerabilities

## Process

The goal is to improve not only the code in this PR but also our skills! The "rules":

- The review is considered "done" as soon as all reviewers have added their review, and all their comments have been addressed.
- For knowledge-sharing reviews, each reviewer should "approve" the PR after studying its content.
- After the approval, the merge is concluded by the developer.

Have fun!

### CODE REVIEW

File: lib/rig/event_filter.ex [hunk 1]
```diff
@@ -270,6 +270,7 @@ defmodule Rig.EventFilter do
 
   """
   @type done_callback :: (() -> nil)
+  @type cloudevent_t :: Cloudevents.t()
```
Comment (kevinbader) [inline lib/rig/event_filter.ex:273]: not an issue, just wondering why... I mean `String.t()` is also never aliased to `string_t`...?

File: lib/rig/event_stream/kafka_to_filter.ex [hunk 2]
```diff
@@ -14,17 +14,17 @@ defmodule Rig.EventStream.KafkaToFilter do
...
-      {:ok, %CloudEvent{} = cloud_event} ->
-        Logger.debug(fn -> inspect(cloud_event.parsed) end)
+  def kafka_handler(body, headers) do
+    case Cloudevents.from_kafka_message(body, headers) do
+      {:ok, cloud_event} ->
+        Logger.debug(fn -> inspect(cloud_event) end)
         EventFilter.forward_event(cloud_event)
         :ok
 
       {:error, :parse_error} ->
```
Comment (kevinbader) [inline lib/rig/event_stream/kafka_to_filter.ex:24]: should be `{:error, reason}`

File: lib/rig/event_stream/kafka_to_http.ex [hunk 3]
```diff
@@ -2,6 +2,8 @@ defmodule Rig.EventStream.KafkaToHttp do
   @moduledoc """
   Forwards all consumed events to an HTTP endpoint.
 
+  TODO: do we want to keep this feature?
```
Comment (kevinbader) [inline lib/rig/event_stream/kafka_to_http.ex:5]: good question... it was only needed this one time, and even then it wasn't a particularly beautiful solution. I guess we can remove it, but probably it makes sense to do this in a dedicated PR.

File: lib/rig/event_filter.ex (hunk 1 above)
Comment (mmacai) [reply lib/rig/event_filter.ex:273]: Because I got this:

`** (CompileError) lib/rig/event_filter.ex:297: cannot invoke remote function Cloudevents.t/0 inside a match`

Comment (kevinbader) [reply lib/rig/event_filter.ex:273]: This error refers to the fact that you cannot match on a _type_ on line 297 - this would only work with a _struct_. Not sure why you don't get the same compiler error with the type alias, cause it should really be exactly the same.
However, there is no way to _match_ for multiple structs at once; instead you'd need to use multiple function clauses for `%Cloudevents.Format.V_1_0.Event{}`, `%Cloudevents.Format.V_0_2.Event`, etc. Not sure that's needed here though.

**Comment by mmacai** [comment]
@kevinbader automated tests are passing, also did some manual testing. I'll do another round of testing in next few days, but you can give it a shot already.

File: lib/rig/event_filter.ex [hunk 4]
```diff
@@ -291,9 +291,9 @@ defmodule Rig.EventFilter do
 
   # ---
 
-  @callback forward_event(CloudEvent.t()) :: :ok
-  @spec forward_event(CloudEvent.t()) :: :ok
-  def forward_event(%CloudEvent{} = event) do
+  @callback forward_event(Cloudevents.t()) :: :ok
+  @spec forward_event(Cloudevents.t()) :: :ok
+  def forward_event(%{} = event) do
```
Comment (kevinbader) [inline lib/rig/event_filter.ex:296]: This is rather un-idiomatic - this is usually done with an `when is_map(event)` guard. However, I'm not sure this is really helpful either way. Haven't tried to write a guard myself so far - is it possible to do something like this:

```elixir
def forward_event(event) when Cloudevents.is_cloudevent(event) do
```

?

File: lib/rig_cloud_events/cloud_event.ex [hunk 5]
```diff
@@ -85,22 +93,40 @@ defmodule RigCloudEvents.CloudEvent do
 
   # ---
 
-  defp specversion_0_1?(parsed) do
+  defp specversion_0_1?(parsed) when is_list(parsed) do
```
Comment (kevinbader) [inline lib/rig_cloud_events/cloud_event.ex:96]: this is .... weird. but only temporary anyway, I assume

File: lib/rig_inbound_gateway/api_proxy/handler/kafka.ex [hunk 6]
```diff
@@ -39,20 +39,25 @@ defmodule RigInboundGateway.ApiProxy.Handler.Kafka do
...
         "HTTP response via Kafka to #{inspect(deserialized_pid)}: #{inspect(message)}"
       end)
 
-      send(deserialized_pid, {:response_received, message})
+      data =
+        if headers == [],
+          do: Cloudevents.Format.Encoder.JSON.encode(event),
+          else: Jason.encode!(event.data)
+
+      send(deserialized_pid, {:response_received, data})
```
Comment (kevinbader) [inline lib/rig_inbound_gateway/api_proxy/handler/kafka.ex:60]: is this only required because we're not using `cloudevents` everywhere yet? perhaps a comment would help future-us..

File: lib/rig_inbound_gateway/api_proxy/handler/kafka.ex [hunk 7]
```diff
@@ -39,20 +39,25 @@ defmodule RigInboundGateway.ApiProxy.Handler.Kafka do
...
          {:ok, correlation_id} <- Map.fetch(rig_metadata, "correlation"),
          {:ok, deserialized_pid} <- Codec.deserialize(correlation_id) do
       Logger.debug(fn ->
         "HTTP response via Kafka to #{inspect(deserialized_pid)}: #{inspect(message)}"
       end)
 
-      send(deserialized_pid, {:response_received, message})
+      data =
+        if headers == [],
+          do: Cloudevents.Format.Encoder.JSON.encode(event),
```
Comment (kevinbader) [inline lib/rig_inbound_gateway/api_proxy/handler/kafka.ex:57]: I think it would be a bit better to use `Cloudevents.to_json(event)` instead

File: lib/rig_inbound_gateway_web/v1/sse.ex [hunk 8]
```diff
@@ -113,7 +113,7 @@ defmodule RigInboundGatewayWeb.V1.SSE do
   end
 
   @impl :cowboy_loop
-  def info(%CloudEvent{} = event, req, state) do
+  def info(%{} = event, req, state) do
```
Comment (kevinbader) [inline lib/rig_inbound_gateway_web/v1/sse.ex:116]: see comment above

File: lib/rig_inbound_gateway_web/v1/sse.ex [hunk 9]
```diff
@@ -197,6 +197,12 @@ defmodule RigInboundGatewayWeb.V1.SSE do
       event: CloudEvent.type!(event)
     }
 
+  defp to_server_sent_event(%{} = event),
```
Comment (kevinbader) [inline lib/rig_inbound_gateway_web/v1/sse.ex:200]: see comment above

File: lib/rig_inbound_gateway_web/v1/sse.ex [hunk 10]
```diff
@@ -197,6 +197,12 @@ defmodule RigInboundGatewayWeb.V1.SSE do
...
+    do: %{
+      data: Cloudevents.Format.Encoder.JSON.encode(event),
```
Comment (kevinbader) [inline lib/rig_inbound_gateway_web/v1/sse.ex:202]: see comment above

File: lib/rig_inbound_gateway_web/v1/websocket.ex [hunk 11]
```diff
@@ -125,7 +125,7 @@ defmodule RigInboundGatewayWeb.V1.Websocket do
   end
 
   @impl :cowboy_websocket
-  def websocket_info(%CloudEvent{} = event, state) do
+  def websocket_info(%{} = event, state) do
```
Comment (kevinbader) [inline lib/rig_inbound_gateway_web/v1/websocket.ex:128]: see comment above

File: CHANGELOG.md [hunk 12]
```diff
@@ -12,7 +12,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Support publishing events consumed from [NATS](https://nats.io) topics. See the [documentation](https://accenture.github.io/reactive-interaction-gateway/docs/event-streams.html#nats) for how to get started. [#297](https://github.com/Accenture/reactive-interaction-gateway/issues/297)
 - Added validation for reverse proxy configuration. Now it crashes RIG on start when configuration is not valid or returns `400` when using REST API to update configuration. [#277](https://github.com/Accenture/reactive-interaction-gateway/issues/277)
 
-<!-- ### Changed -->
+### Changed
+
+- Incorporated [cloudevents-ex](https://github.com/kevinbader/cloudevents-ex) to handle binary and structured modes for [Kafka protocol binding](https://github.com/cloudevents/spec/blob/v1.0/kafka-protocol-binding.md) in a proper way. This introduces some **breaking changes**:
+  - Binary mode is now using `ce_` prefix for CloudEvents context attribute headers, before it was `ce-`
+  - `KAFKA_SCHEMA_REGISTRY_HOST` env var now expects also http protocol, e.g. - `http://localhost:8081`, before it was `localhost:8081`
```
[the remark this answers is not in the document]
Comment (mmacai) [reply CHANGELOG.md:19]: I guess, but ideally do that in cloudevents lib since it's caused by the Avrora lib.

File: docs/avro.md [hunk 13]
```diff
@@ -63,25 +63,25 @@ docker run --name rig \
...
+21:54:52.869 module=Avrora.Storage.Registry [debug] obtaining schema with global id `1`
+21:54:52.870 module=Rig.EventStream.KafkaToFilter [debug] %Cloudevents.Format.V_0_2.Event{contenttype: "avro/binary", data: %{"example" => "test"}, extensions: %{"rig" => %{"correlation" => "Ve1d-XF0Qi46lwh47X5IqI7m_FCIqCLsqyV0KTCxg28Hnd7ytczBe1cASZYPxA7GNFCZ4AzDC0QX1w0=", "headers" => [["accept", "*/*"], ["content-length", "221"], ["content-type", "application/json"], ["host", "localhost:4000"], ["user-agent", "curl/7.54.0"]], "host" => "localhost", "method" => "POST", "path" => "/myapi/publish-async", "port" => 4000, "query" => "", "remoteip" => "172.28.0.1", "scheme" => "http"}}, id: "069711bf-3946-4661-984f-c667657b8d85", schemaurl: nil, source: "/cli", specversion: "0.2", time: "2018-04-05T17:31:00Z", type: "com.example"}
 ```
 
 ## Example 2: Kafka schema Registry CLI
 
 To check if it works also with native serializer we can leverage the CLI shipped with the Kafka Schema Registry image.
 
-```bash
+``` bash
```
[the remark this answers is not in the document]
Comment (mmacai) [reply docs/avro.md:84]: It should, my auto-formatter just did just "magic".

File: lib/rig_cloud_events/cloud_event.ex (hunk 5 above)
Comment (mmacai) [reply lib/rig_cloud_events/cloud_event.ex:96]: Yes, temporary, we should do another PRs to migrate also other stuff to use the couldevents lib.

File: lib/rig_inbound_gateway/api_proxy/handler/kafka.ex (hunk 6 above)
Comment (mmacai) [reply lib/rig_inbound_gateway/api_proxy/handler/kafka.ex:60]: Exactly, should be gone once we migrate everything to use the cloudevents lib. Would do it in separate PRs though.

File: lib/rig/event_filter.ex (hunk 4 above)
Comment (mmacai) [reply lib/rig/event_filter.ex:296]: Got

```bash
== Compilation error in file lib/rig/event_filter.ex ==
** (CompileError) lib/rig/event_filter.ex:296: cannot invoke remote function Cloudevents.is_cloudevent/1 inside guards
```

File: lib/rig_inbound_gateway/api_proxy/handler/kafka.ex (hunk 6 above)
Comment (mmacai) [reply lib/rig_inbound_gateway/api_proxy/handler/kafka.ex:60]: Wait, what do you mean? Sorry, thought it's a different part of the code.

File: CHANGELOG.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (mmacai) [reply CHANGELOG.md:19]: https://github.com/kevinbader/cloudevents-ex/pull/4

File: docs/avro.md [hunk 14]
```diff
@@ -23,38 +23,38 @@ Adopting Avro for event (de)serialization is fairly straightforward. First you n
...
-- If it is, creates headers for Kafka event by appending `ce-` prefix for every field, besides `data` field
-  - **nested context attributes are stringified**, since Kafka headers don't support nested values (this is common when using Cloud events extensions)
-- after that, the `data` field is serialized using the schema name (function for getting schemas from registry is cached in-memory)
-- producer sends event with created headers and data (in binary format `<<0, 0, 0, 0, 1, 5, 3, 8, ...>>`) to Kafka
+* producer evaluates if serialization is turned on by checking `KAFKA_SERIALIZER` environment variable and if it's value is `avro`
+* If it is, creates headers for Kafka event by appending `ce_` prefix for every field, except `data` field - **binary mode**
+  + **nested context attributes are stringified**, since Kafka headers don't support nested values (this is common when using Cloud events extensions)
+* after that, the `data` field is serialized using the schema name (function for getting schemas from registry is cached in-memory)
+* producer sends event with created headers and data (in binary format `<<0, 0, 0, 0, 1, 5, 3, 8, ...>>` ) to Kafka
```
Comment (kevinbader) [inline docs/avro.md:30]: I think your formatter is crap ^_^

File: lib/rig/event_stream/kafka_to_filter.ex [hunk 15]
```diff
@@ -6,25 +6,24 @@ defmodule Rig.EventStream.KafkaToFilter do
...
+    case Cloudevents.from_kafka_message(body, headers) do
+      {:ok, cloud_event} ->
+        Logger.debug(fn -> inspect(cloud_event) end)
         EventFilter.forward_event(cloud_event)
         :ok
 
-      {:error, :parse_error} ->
-        {:error, :non_cloud_events_not_supported, message}
+      {:error, _reason} ->
+        {:error, :non_cloud_events_not_supported, body}
```
Comment (kevinbader) [inline lib/rig/event_stream/kafka_to_filter.ex:24]: including the `reason` in the `:error` tuple would make it visible in logs, so I suggest changing this to

```elixir
{:error, {:non_cloud_events_not_supported, reason, body}}
```

File: lib/rig/event_stream/kafka_to_filter.ex [hunk 16]
```diff
@@ -6,25 +6,24 @@ defmodule Rig.EventStream.KafkaToFilter do
...
     end
   rescue
-    err -> {:error, err, message}
+    err -> {:error, err, body}
```
Comment (kevinbader) [inline lib/rig/event_stream/kafka_to_filter.ex:27]: to make this similar to the other error you could do instead:

```elixir
{:error, {:failed_to_parse_kafka_message, headers, body}}
```

**Comment by mmacai** [comment]
@kevinbader @Knappek merged with master tracing changes and got some complications here. Took an easy/temporary approach to make it. Automated and manual tests work, but no idea about tracing, can you check that?
